src/api/system.c

```diff
@@ -109,9 +109,14 @@ static const char *get_key_name(const SDL_Event *e, char *buf) {
     /* We need to correctly handle non-standard layouts such as dvorak.
        Therefore, if a Latin letter(code<128) is pressed in the current layout,
        then we transmit it as it is. But we also need to support shortcuts in
-       other languages, so for non-Latin characters we pass the scancode that
-       matches the letter in the QWERTY layout. */
-    if (e->key.keysym.sym < 128)
+       other languages, so for non-Latin characters(code>128) we pass the
+       scancode based name that matches the letter in the QWERTY layout.
+       
+       In SDL, the codes of all special buttons such as control, shift, arrows
+       and others, are shifted using bit shift 1<<30(SDLK_SCANCODE_MASK) outside
```

Comment:
It looks like SDL doesn't bit shift the values, it just `or`s with the mask.

Reply:
...are shifted using bit mask 1<<30...?

Reply:
The codes are not shifted. The shift only applies to the `1`.
The code they use:
```C
#define SDLK_SCANCODE_MASK (1<<30)
```
becomes
```C
#define SDLK_SCANCODE_MASK 0b01000000000000000000000000000000
```
so that this
```C
#define SDL_SCANCODE_TO_KEYCODE(X)  (X | SDLK_SCANCODE_MASK)
```
basically just does `X | 0b01000000000000000000000000000000`.
So it's just masking the value.

Reply:
Moved? Doing this is like adding 2^30. Therefore, initially there was a comparison.

Reply:
Doing the comparison like you did is fine in this context.
My problem is with the comment, as it doesn't represent correctly what SDL is doing.

What you wrote in the comment makes it look like SDL shifts the codes, which is not correct.
What SDL does, is ORing the code with `SDLK_SCANCODE_MASK`.

Reply:
There is a set of special buttons, the scancodes of which are in the range from 0 to 512. Sdl, when converting to keycodes, shifts this set of scancodes from 0 to 2^30. This is the same as saying that it changes the 30th bit. Or that a bitwise OR operation with a mask of 1<<30 is applied. I don't see the difference between all of this.
Your English is better than mine. Write whatever you want. You seem to have rights.

Reply:
> Sdl, when converting to keycodes, shifts this set of scancodes from 0 to 2^30. This is the same as saying that it changes the 30th bit.

I think the problem here is that your definition of shifting is incorrect. Shifting means inserting zeros at the end of a binary number; what you are referring to is OR masking.
So, the shifting part is _only_ `1<<30`, while setting the 30th bit is the masking part.

I'd write the comment as `In SDL, the codes of all special buttons such as control, shift, arrows and others, are masked with SDLK_SCANCODE_MASK, which moves them outside the unicode range (>0x10FFFF).`